### DISCUSSION TITLE
How to know the port of a message in a callback ?

### DISCUSSION THREAD

**Discussion by mikacousin** [mido/mido#389] [Q&A] [answered]
Hello,

I listen to several MIDI controllers with a single callback function (the number of controllers is not fixed).
My problem is that I need to know the controller that sends the received MIDI message to adapt the modes used (Makie, absolute, ...).
Actually, I'll need to add the port to the callback parameters in addition to the MIDI message, but I don't know how to do that.
I hope to be understandable :-)

Thank you very much for this very good project!

**Comment by rdoursenaud** [comment]
Callbacks are a per-port setting. Just pass a different function for each port.

**Comment by mikacousin** [comment]
Thank you for your answer.
(I still have a lot to learn about python)
It was my first idea (different function for each port), but the number of controllers can vary enormously per user, I don't know how to code this in general (undetermined number of functions).
Maybe a solution with Inheritance, but I have to study that ...

I'll keep looking, I'll post my solution when I find it.

**Reply by rdoursenaud** [reply]
If you need a generic solution, one way to go would be to create your own class to encapsulate the midi port and callback. I've done something similar in [MIDI Explorer](https://github.com/ematech/midiexplorer).

Something like:
```python
import mido

class MyMidiIn():
    """A thin wrapper around mido input port"""
    name: str  # The port name
    port: mido.ports.BaseInput   # The port itself

    def __init__(self, name: str) -> None:
        self.name = name
        self.port = mido.open_input(self.name)
        self.port.callback = self.receive_callback

    def __del__(self) -> None:
       self.port.callback = None
       self.port.close()

    def receive_callback(self, message: mido.Message) -> None:
        """A generic callback"""
        # Do your thing here. You have access to self.name and self.port to know which port you are.

###
# Usage
###
my_input = MyMidiIn("the input name")  # Opens the port and sets the callback
```

There are other methods, this is just the first that came to mind.

**Reply by mikacousin** [reply]
Thank you so much !
Your solution is infinitely more elegant.
I still have a lot to learn :-)

**Reply by williamcorney** [reply]
I have tried the code above and something similar but its not working for me and I'm not sure where I am going wrong.

I am probably not understanding a fundamental concept of how the program flow works with callbacks , classes etc.

**What I am trying to do?**

I want to listen to midi input and send it to a callback handler which contains a series of logic statements.

**What happens?**

Script runs without error but is exiting immediately rather than as it were staying and doing whatever is defined in note handler eg printing note details to screen

**Example Script**
```
import mido
portname = "Digital Piano"
def note_handler(note: mido.Message) -> None:
   print ('blah blah')


with mido.open_input(portname, callback=note_handler) as port:
     print("Using {}".format(port))

```

**Reply by williamcorney** [reply]
Ok so after much trial and error this seems to work.  The only difference I can see is that you seem to need to have some kind of loop open , although, if u understand correctly this doesn't necessarily have to be a time.sleep, this might be achieved if this script was integrated into a gui application which had a mainloop function??



```

import mido
import time
def listen_input(message):
    print('input:', message)

inport = mido.open_input('Digital Piano', callback=listen_input)

while True:
    time.sleep(1)

```

**Reply by rdoursenaud** [reply]
Sorry for the late reply. The callback is asynchronously called in another thread so, yes, you need some kind of mainloop in your program to keep it running. In MidiExplorer, I effectively use the GUI’s mainloop.

**Comment by mikacousin** [comment]
I think I found a solution by creating callback functions dynamically (thanks to stackoverflow.com).

Here is a sample script as proof of concept (interested in your opinions) :
```py
import mido


def function_builder(port):
    def scan(msg):
        print(port, msg)
    return scan


inports = list(set(mido.get_input_names()))
scan_functions = {}

for port in inports:
    print(port)
    inport = mido.open_input(port)
    scan_functions[port] = function_builder(port)
    inport.callback = scan_functions[port]

while True:
    pass
```

**Reply by rdoursenaud** [reply]
This looks clunky to me. Check my answer above.

**Comment by mikacousin** [comment, marked as answer]
@rdoursenaud's code (see https://github.com/mido/mido/discussions/389#discussioncomment-3766289) does the job. Thanks!
